File: WebClient SDK/src/default/ui/krProcess/stageHandlers/signingUIHandler.ts
import { DefaultCardTypes, plainColumnName } from 'tessa/workflow';
import {
  IKrStageTypeUIHandlerContext,
  KrStageTypeUIHandler,
  StageTypeHandlerDescriptor,
  signingDescriptor
} from 'tessa/workflow/krProcess';

import { CardFieldChangedEventArgs } from 'tessa/cards';
import { IControlViewModel } from 'tessa/ui/cards';
import { TabControlViewModel } from 'tessa/ui/cards/controls';
import { Visibility } from 'tessa/platform';

/**
 * UI обработчик типа этапа {@link signingDescriptor}.
 */
export class SigningUIHandler extends KrStageTypeUIHandler {
  //#region fields

  private static readonly notReturnEdit = plainColumnName(
    'KrSigningStageSettingsVirtual',
    'NotReturnEdit'
  );

  private _returnIfNotSignedFlagControl?: IControlViewModel;
  private _returnAfterSigningFlagControl?: IControlViewModel;

  //#endregion

  //#region base overrides

  public descriptors(): StageTypeHandlerDescriptor[] {
    return [signingDescriptor];
  }

  public async initialize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    let flagsTabs: IControlViewModel | undefined;

    if (
      (flagsTabs = context.settingsForms
        .find(i => i.name === DefaultCardTypes.KrSigningStageTypeSettingsTypeName)
        ?.blocks.find(i => i.name === 'SigningStageFlags')
        ?.controls.find(x => x.name === 'FlagsTabs')) &&
      flagsTabs instanceof TabControlViewModel
    ) {
      this._returnIfNotSignedFlagControl = flagsTabs.tabs
        .find(x => x.name === 'CommonSettings')
        ?.blocks.find(x => x.name === 'StageFlags')
        ?.controls.find(x => x.name === 'ReturnIfNotSigned');

      this._returnAfterSigningFlagControl = flagsTabs.tabs
        .find(x => x.name === 'AdditionalSettings')
        ?.blocks.find(x => x.name === 'StageFlags')
        ?.controls.find(x => x.name === 'ReturnAfterSigning');
    }

    context.row.fieldChanged.add(this.onSettingsFieldChanged);

    this.notReturnEditConfigureFields(context.row.tryGet(SigningUIHandler.notReturnEdit, false));
  }

  public async finalize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    context.row.fieldChanged.remove(this.onSettingsFieldChanged);
  }

  //#endregion

  //#region private methods

  private onSettingsFieldChanged = (e: CardFieldChangedEventArgs) => {
    if (e.fieldName === SigningUIHandler.notReturnEdit) {
      this.notReturnEditConfigureFields(e.fieldValue);
    }
  };

  private notReturnEditConfigureFields(isNotReturnEdit: boolean) {
    if (isNotReturnEdit) {
      if (this._returnIfNotSignedFlagControl) {
        this._returnIfNotSignedFlagControl.controlVisibility = Visibility.Collapsed;
      }

      if (this._returnAfterSigningFlagControl) {
        this._returnAfterSigningFlagControl.controlVisibility = Visibility.Collapsed;
      }
    } else {
      if (this._returnIfNotSignedFlagControl) {
        this._returnIfNotSignedFlagControl.controlVisibility = Visibility.Visible;
      }

      if (this._returnAfterSigningFlagControl) {
        this._returnAfterSigningFlagControl.controlVisibility = Visibility.Visible;
      }
    }
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/StageHandlers/SigningUIHandler.cs
#nullable enable

using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.UI.KrProcess.StageHandlers
{
    /// <summary>
    /// UI обработчик типа этапа <see cref="StageTypeDescriptors.SigningDescriptor"/>.
    /// </summary>
    public sealed class SigningUIHandler :
        StageTypeUIHandlerBase
    {
        #region Fields

        private IControlViewModel? returnIfNotSignedFlagControl;
        private IControlViewModel? returnAfterSigningFlagControl;

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override Task Initialize(IKrStageTypeUIHandlerContext context)
        {
            if (context.SettingsForms.FirstOrDefault(static i => i.Name == DefaultCardTypes.KrSigningStageTypeSettingsTypeName) is { } form
                && form.Blocks.FirstOrDefault(static i => i.Name == "SigningStageFlags") is { } flagsBlock
                && flagsBlock.Controls.FirstOrDefault(static i => i.Name == "FlagsTabs") is TabControlViewModel flagsTabsViewModel)
            {
                this.returnIfNotSignedFlagControl = flagsTabsViewModel
                    .Tabs.FirstOrDefault(static i => i.Name == "CommonSettings")
                    ?.Blocks.FirstOrDefault(static i => i.Name == "StageFlags")
                    ?.Controls.FirstOrDefault(static i => i.Name == KrConstants.Ui.ReturnIfNotSigned);

                this.returnAfterSigningFlagControl = flagsTabsViewModel
                    .Tabs.FirstOrDefault(static i => i.Name == "AdditionalSettings")
                    ?.Blocks.FirstOrDefault(static i => i.Name == "StageFlags")
                    ?.Controls.FirstOrDefault(static i => i.Name == KrConstants.Ui.ReturnAfterSigning);
            }

            context.Row.FieldChanged += this.OnSettingsFieldChanged;

            this.NotReturnEditConfigureFields(context.Row.TryGet<bool>(KrConstants.KrSigningStageSettingsVirtual.NotReturnEdit));

            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public override Task Finalize(IKrStageTypeUIHandlerContext context)
        {
            context.Row.FieldChanged -= this.OnSettingsFieldChanged;

            return Task.CompletedTask;
        }

        #endregion

        #region Private Methods

        /// <summary>
        /// Обработчик события <see cref="CardRow.FieldChanged"/> строки настроек этапа.
        /// </summary>
        /// <param name="sender">Источник события.</param>
        /// <param name="e">Информация о событии.</param>
        private void OnSettingsFieldChanged(object? sender, CardFieldChangedEventArgs e)
        {
            if (e.FieldName == KrConstants.KrSigningStageSettingsVirtual.NotReturnEdit
                && e.FieldValue is bool notReturnEdit)
            {
                this.NotReturnEditConfigureFields(notReturnEdit);
            }
        }

        /// <summary>
        /// Настраивает поля в зависимости от значения флага <see cref="KrConstants.KrSigningStageSettingsVirtual.NotReturnEdit"/>.
        /// </summary>
        /// <param name="isNotReturnEdit">Значение флага <see cref="KrConstants.KrSigningStageSettingsVirtual.NotReturnEdit"/>.</param>
        private void NotReturnEditConfigureFields(bool isNotReturnEdit)
        {
            if (isNotReturnEdit)
            {
                if (this.returnIfNotSignedFlagControl is not null)
                {
                    this.returnIfNotSignedFlagControl.ControlVisibility = Visibility.Collapsed;
                }

                if (this.returnAfterSigningFlagControl is not null)
                {
                    this.returnAfterSigningFlagControl.ControlVisibility = Visibility.Collapsed;
                }
            }
            else
            {
                if (this.returnIfNotSignedFlagControl is not null)
                {
                    this.returnIfNotSignedFlagControl.ControlVisibility = Visibility.Visible;
                }

                if (this.returnAfterSigningFlagControl is not null)
                {
                    this.returnAfterSigningFlagControl.ControlVisibility = Visibility.Visible;
                }
            }
        }

        #endregion
    }
}
